In [2]:
# import libraries
import pandas as pd
import numpy as np
import os

In [3]:
# establish path to project folder
path = r'C:\Users\BrianDavis\OneDrive - abetterchicago.org\Documents\Data Analysis\Achievement 5 (Python)\04-2023 Instacart Basket Analysis'

In [4]:
# import "products" file
df_prods = pd.read_csv(os.path.join(path,'02 Data','Original Data','products.csv'), index_col = False)

In [5]:
# import "orders wrangled" file
df_ords = pd.read_csv(os.path.join(path,'02 Data','Prepared Data','orders_wrangled.csv'), index_col = False)

In [6]:
df_ords = df_ords.drop(columns = 'Unnamed: 0')

# Exercise 5.5 

In [7]:
#create a dataframe
df_test = pd.DataFrame()

In [8]:
#create a mixed type column
df_test['mix'] = ['a', 'b', 1, True]

In [9]:
df_test

mix
0     a
1     b
2     1
3  True

In [10]:
#check for mixed types 

for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [11]:
#change datatypes
df_test['mix'] = df_test['mix'].astype('str')

In [12]:
#count missing values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [13]:
#create variable with info on rows with missing product names
df_nan = df_prods.loc[df_prods['product_name'].isnull() == True]

In [14]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [15]:
df_prods.shape

(49693, 5)

In [16]:
#create new dataframe that excludes missing product nanmes 
df_prods_clean = df_prods.loc[df_prods['product_name'].isnull() == False]

In [17]:
df_prods_clean.shape

(49677, 5)

In [18]:
#update existing dataframe to exclude missing product nanmes 

df_prods.dropna(subset = ['product_name'], inplace = True)

In [19]:
df_prods.shape

(49677, 5)

In [20]:
#look for duplicate values 
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [21]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [22]:
df_prods_clean.shape

(49677, 5)

### Check for outliers in products file

In [23]:
df_prods_clean.describe()

product_id      aisle_id  department_id        prices
count  49677.000000  49677.000000   49677.000000  49677.000000
mean   24850.194235     67.763110      11.728687      9.993164
std    14340.588602     38.316396       5.850651    453.592708
min        1.000000      1.000000       1.000000      1.000000
25%    12433.000000     35.000000       7.000000      4.100000
50%    24851.000000     69.000000      13.000000      7.100000
75%    37267.000000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000  99999.000000

#### Observation: the max price value of 99,999 is an outlier and should be excluded. 

In [24]:
# check for products that cost more than 50 
df_prods_clean[df_prods_clean['prices'] > 50]

product_id                      product_name  aisle_id  department_id  \
21554       21553  Lowfat 2% Milkfat Cottage Cheese       108             16   
33666       33664             2 % Reduced Fat  Milk        84             16   

        prices  
21554  14900.0  
33666  99999.0

#### Observation: there are two products that are priced over 50; these appear to be inputted incorrectly (multipled by 10,000)

In [25]:
#replace outlier values in products file 
df_prods_clean = df_prods_clean.replace ({'prices': {99999.0 : 9.99, 14900.0 : 1.49}})

In [26]:
#check that changes were made correctly 
df_prods_clean.describe()

product_id      aisle_id  department_id        prices
count  49677.000000  49677.000000   49677.000000  49677.000000
mean   24850.194235     67.763110      11.728687      7.680473
std    14340.588602     38.316396       5.850651      4.199307
min        1.000000      1.000000       1.000000      1.000000
25%    12433.000000     35.000000       7.000000      4.100000
50%    24851.000000     69.000000      13.000000      7.100000
75%    37267.000000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000     25.000000

In [27]:
#delete duplicate rows
df_prods_clean_no_dup = df_prods_clean.drop_duplicates()

# Assignment 5.5

## 2. Run the df.describe() function on your df_ords dataframe. Using your new knowledge about how to interpret the output of this function, share in a markdown cell whether anything about the data looks off or should be investigated further.


In [28]:
#describe orders dataframe
df_ords.describe()

order_id       user_id  order_number  order_day_of_week  \
count  3.421083e+06  3.421083e+06  3.421083e+06       3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01       2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01       2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00       0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00       1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01       3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01       5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02       6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

#### Observation: The data looks good. The ranges for each column all make sense, and the other statistics are logical.

## 3. Check for mixed-type data in your df_ords dataframe.

## 4. If you find mixed-type data, fix it. The column in question should contain observations of a single data type.


In [29]:
#check for mixed types 

for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

In [30]:
#### Observation: no mixed type data is found in the orders file

## 5. Run a check for missing values in your df_ords dataframe.


In [31]:
#call information on orders file
df_ords.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3421083 entries, 0 to 3421082
Data columns (total 6 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   order_id                int64  
 1   user_id                 int64  
 2   order_number            int64  
 3   order_day_of_week       int64  
 4   order_hour_of_day       int64  
 5   days_since_prior_order  float64
dtypes: float64(1), int64(5)
memory usage: 156.6 MB


In [32]:
#check for missing values

df_ords.isnull().sum()

order_id                       0
user_id                        0
order_number                   0
order_day_of_week              0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

#### Observation: there are 206,209 missing values in the "days_since_prior_order" column; because all other logical values are present, these N/A's could potentially be user/system errors

## Address the missing values using an appropriate method.

#### Note: I chose not to exclude these values because they make up more than 10% of the values in the entire dataframe and doing so would materially alter future analyses based on this data. 

## Run a check for duplicate values in your df_ords data.

In [33]:
#check for duplicate values by creating new variable

df_ords_dups = df_ords[df_ords.duplicated()]

In [34]:
#call this variable to check for duplicate values
df_ords_dups

Empty DataFrame
Columns: [order_id, user_id, order_number, order_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

#### Observation: there are no duplicate values in the orders file 

# Export Dataframes

In [35]:
#export products file 
df_prods_clean_no_dup.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_clean_wrangled.csv'))

In [37]:
#export orders file 
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_clean_wrangled.csv'))